cdap-app-templates/cdap-etl/cdap-etl-realtime/src/main/java/co/cask/cdap/templates/etl/realtime/ETLWorker.java

```diff
@@ -0,0 +1,207 @@
+
+  private void initializeSink(WorkerContext context, ETLStage stage) throws Exception {
+    StageSpecification spec = GSON.fromJson(context.getRuntimeArguments().get(Constants.Sink.SPECIFICATION),
+                                            StageSpecification.class);
+    sink = (RealtimeSink) Class.forName(spec.getClassName()).newInstance();
+    WorkerSinkContext sinkContext = new WorkerSinkContext(context, stage, spec, datasetContext);
+    LOG.info("Sink Stage : {}", stage.getName());
+    LOG.info("Sink Class : {}", stage.getClass().getName());
+    LOG.info("Specifications of Sink : {}", spec);
+    sink.initialize(sinkContext);
+  }
+
+  private void initializeTransforms(WorkerContext context, List<ETLStage> stages) {
+    List<StageSpecification> specs = GSON.fromJson(context.getRuntimeArguments().get(
+      Constants.Transform.SPECIFICATIONS), SPEC_LIST_TYPE);
+    for (int i = 0; i < specs.size(); i++) {
+      StageSpecification spec = specs.get(i);
+      ETLStage stage = stages.get(i);
+      try {
+        Transform transform = (Transform) Class.forName(spec.getClassName()).newInstance();
+        DefaultTransformContext transformContext = new DefaultTransformContext(spec, stage.getProperties());
+        transform.initialize(transformContext);
+        transforms.add(transform);
+      } catch (ClassNotFoundException e) {
+        LOG.error("Unable to load Transform : {}", spec.getClassName(), e);
+        Throwables.propagate(e);
+      } catch (InstantiationException e) {
+        LOG.error("Unable to instantiate Transform : {}", spec.getClassName(), e);
+        Throwables.propagate(e);
+      } catch (IllegalAccessException e) {
+        LOG.error("Error while creating instance of Transform : {}", spec.getClassName(), e);
+        Throwables.propagate(e);
+      }
+    }
+  }
+
+  @Override
+  public void run() {
+    running = true;
+    while (running) {
+      final SourceState sourceState = new SourceState();
+
+      // Fetch SourceState from State Table.
+      getContext().execute(new TxRunnable() {
+        @Override
+        public void run(DatasetContext context) throws Exception {
+          KeyValueTable stateTable = context.getDataset(ETLRealtimeTemplate.STATE_TABLE);
+          byte[] stateBytes = stateTable.read(adapterName);
+          if (stateBytes != null) {
+            SourceState state = GSON.fromJson(Bytes.toString(stateBytes), SourceState.class);
+            sourceState.setState(state.getState());
+          }
+        }
+      });
+
+      final SourceState nextState = source.poll(defaultEmitter, sourceState);
+      for (Object sourceData : defaultEmitter) {
+        try {
+          final Iterable<Object> dataToSink = transformExecutor.runOneIteration(sourceData);
+          getContext().execute(new TxRunnable() {
+            @Override
+            public void run(DatasetContext context) throws Exception {
+              for (Object object : dataToSink) {
+                sink.write(object);
```

Comment:
hm do you think it makes sense to pass the iterable in to the sink? Some sinks may be able to write more efficiently that way